### ISSUE TITLE
Error in new.session() 

### ISSUE THREAD

**Issue by abernal30** [joshuaulrich/quantmod#358]
### Description

When applying the function `getSymbols()`, it errors and does not return a result.

### Minimal, reproducible example

```r
getSymbols("AAPL")
## Error in new.session() : Could not establish session after 5 attempts.
```

**Comment by pverspeelt** [comment]
@joshuaulrich,

The issue is located in the function `.getHandle` and specifically in the local function `new.session`. All the 5 calls to the yahoo website return with no cookies, hence the function fails. 

Looking over the code. I don't see much point in the `.gethandle` function as the yahoo query is also being built in the `.yahooURL` function. The only thing that `.gethandle` gives to `.yahooURL` is the crumb request to the query.srv. and cb has the value: "" for either query1 or query2. So this part is not needed to get the data. 

Removing the `.gethandle` and removing the crumb part from building the yahoo url in `.yahooURL` returns the data without any issues.

**Comment by mpadge** [comment]
seeing same issue here

**Comment by alihru** [comment]
one also needs to get rid of the if statement inside the .getHandle function which stops the process if the number of cookies is zero 

if (NROW(curl::handle_cookies(h)) ==0
 stop(„Could not establish session after 5 attempts.“)

p.s. thank you for the quick fix! Really appreciate it

**Comment by asaeb1480** [comment]
what's the solution? These comments are not working!

**Comment by datawookie** [comment]
@asaeb1480 please try installing from the `358-getsymbols-new.session` branch. It's working for me again.

**Comment by asaeb1480** [comment]
What's the exact code for installation of 358-getsymbols-new.session plz?

**Comment by datawookie** [comment]
@asaeb1480 please look at https://cran.r-project.org/web/packages/githubinstall/vignettes/githubinstall.html.

**Comment by joshuaulrich** [comment]
`remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")`

**Comment by m1270** [comment]
Thank you @joshuaulrich 

I can't install it due to this error: 

Error: package ‘xts’ 0.12.1 was found, but > 0.12.1 is required by ‘quantmod’

The latest version of xts package is 0.12.1. I am running ubuntu 20.04 and R version of 4.1.2

**Comment by joshuaulrich** [comment]
@m1270 ah, that's because I have a change in development to move the S4 xts/zoo class to xts... I need to revert that. Thanks for the catch!

**Comment by fededemo** [comment]
Thanks joshua. getSymbols working now. 
@m1270 working for me with an remotes::install_github("joshuaulrich/xts", force = TRUE)

**Comment by mik3hall** [comment]
That installs but doesn't appear to fix for me.

**Comment by TommyTorstP** [comment]
I have the same issue as @mik3hall here! I tried to find it in the given directory on the computer, but the file still seems to be deleted once install is complete

**Comment by asaeb1480** [comment]
Thanks, Joshuaulrich. It's working for me.

**Comment by mik3hall** [comment]
@TommyTorstP it is in the user library path that it indicated in the install but I must not default search that. Looking at how I can include that now. I assume it will eventually cran install to a normal system path.

**Comment by pawatkins** [comment]
Hello,

I'm very sorry but I'm an idiot when it comes to this stuff. I'm currently enrolled in a master's class but I've only been coding since January 2022.

I have a project due tonight at midnight and it uses quantmod. 

How can I fix this error?  I'm in Rstudio

Thanks in advance. I know this has to be work.

**Comment by Baroni77** [comment]
after a bit of experimenting, it is working for me too. Thanks @joshuaulrich!

**Comment by m1270** [comment]
Thank you @joshuaulrich 
I confirm it is working for me as well.

**Comment by MiraiLi00** [comment]
Hello, @joshuaulrich, thank you for helping with this issue, but I was trying to run the code you provided, but it is still not working for me and it gives me this error:

ERROR: package installation failed
Error: Failed to install 'quantmod' from GitHub:
System command 'Rcmd.exe' failed, exit status: 1, stdout & stderr were printed

(sorry I am new to this)

Do you know what could be a problem?

**Comment by eddelbuettel** [comment]
@joshuaulrich methinks you still have the wrong `xts` constraint for general installations:

```sh
edd@rob:~$ Rscript -e 'remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")'
Using github PAT from envvar GITHUB_PAT
Downloading GitHub repo joshuaulrich/quantmod@358-getsymbols-new.session

✔  checking for file ‘/tmp/RtmpI0Q3cp/remotes1ebcff64cdb8cc/joshuaulrich-quantmod-0a90c3a/DESCRIPTION’ ...
─  preparing ‘quantmod’:
✔  checking DESCRIPTION meta-information ...
─  installing the package to process help pages
         -----------------------------------
─  installing *source* package ‘quantmod’ ...
   ** using staged installation
   ** R
   ** demo
   ** byte-compile and prepare package for lazy loading
   Error: package ‘xts’ 0.12.1 was found, but > 0.12.1 is required by ‘quantmod’
   Execution halted
   ERROR: lazy loading failed for package ‘quantmod’
─  removing ‘/tmp/Rtmp1v5YAX/Rinst1ebdff5823613b/quantmod’
         -----------------------------------
   ERROR: package installation failed
Error: Failed to install 'quantmod' from GitHub:
  System command 'R' failed, exit status: 1, stdout & stderr were printed
Execution halted
edd@rob:~$ 
```

**Comment by joshuaulrich** [comment]
Ugh, I made the commit but didn't push. Thanks @eddelbuettel.

**Comment by awp-finanznachrichten** [comment]
Hello,

Since we are also working with your package, we are very happy about the quick fixing of the bug, thank you!

Solution of @fededemo for xts was working for us, too:
1. remotes::install_github("joshuaulrich/xts", force = TRUE)
2.  remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")

**Comment by eddelbuettel** [comment]
@awp-finanznachrichten Please review and then edit your post.  Step 1 is _now_ no longer needed.  You were too fast.

**Comment by abernal30** [comment]
@joshuaulrich  It works pretty fine now, thank so much for your time and effort!!!

**Comment by pyrfin** [comment]
@joshuaulrich It works now. Thanks for your support!

**Comment by OmegaSquared** [comment]
Following worked for me: 

`remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")`

Thank you @joshuaulrich  for having a solution so quickly. This is why I love your work, when something breaks you are on top of it. Cheers mate.

**Comment by shumakerro** [comment]
Thanks Josh, got it working and I'll echo all the compliments for your work now and always.

**Comment by cynthianahartono** [comment]
Does using remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session") mean the getSymbol function cannot be used?

Been trying to use it but got this error "could not find function "getSymbols""

**Comment by samiranvekar** [comment]
post executing remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")  with option None

and then typing library(quantmod) I get the following error

Loading required package: xts
Loading required package: zoo

Attaching package: ‘zoo’

The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric

Loading required package: TTR
Error in value[[3L]](cond) : 
  Package ‘quantmod’ version 0.4.18 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘quantmod’ is imported by ‘FinancialInstrument’, ‘quantstrat’, ‘blotter’ so cannot be unloaded
In addition: Warning messages:
1: package ‘xts’ was built under R version 4.0.5 
2: package ‘TTR’ was built under R version 4.0.5 

Hence I am not able to use getSymbol function

How to resolve this?

**Comment by TH3Heurist** [comment]
@joshuaulrich You are a legend! Thank you sir.

**Comment by localperf** [comment]
remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")

works for me!  Thanks!

**Comment by MiraiLi00** [comment]
Now it is working for me too! Thank you very much to @joshuaulrich for resolving the issue and also to all of you for your useful comments!!

**Comment by msperlin** [comment]
Same here (linux mint 20.3). Thanks!

**Comment by joshuaulrich** [comment]
> Does using remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session") mean the getSymbol function cannot be used?
> 
> Been trying to use it but got this error "could not find function "getSymbols""

@cynthianahartono looks like you need to run `library("quantmod")`. You need that command before you can use `getSymbols()`.

**Comment by joshuaulrich** [comment]
> Package ‘quantmod’ version 0.4.18 cannot be unloaded:
> Error in unloadNamespace(package) : namespace ‘quantmod’ is imported by ‘FinancialInstrument’, ‘quantstrat’, ‘blotter’ so cannot be unloaded

@samiranvekar You probably need to restart R and then install the new quantmod before you load those other packages that depend on it.

**Comment by samiranvekar** [comment]
Thanks@joshuaulrich. I will try the suggestion

However when the permanent fix will be delivered for quantmod package in CRAN

SamirA

**Comment by cynthianahartono** [comment]
> > Does using remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session") mean the getSymbol function cannot be used?
> > Been trying to use it but got this error "could not find function "getSymbols""
> 
> @cynthianahartono looks like you need to run `library("quantmod")`. You need that command before you can use `getSymbols()`.

@joshuaulrich works like a charm, thank you!!

**Comment by pkinif** [comment]
Thanks for solving this issue so quickly @joshuaulrich

**Comment by BathosEconTrade** [comment]
Man, had no idea how dependent I had become on Quandl. 
Thank you for your help and look forward to the CRAN fix.  Any idea when?

**Comment by joshuaulrich** [comment]
@samiranvekar and @BathosEconTrade, I hope to have it on CRAN later today. I'm still working through some reverse dependency checks. Keep in mind that it usually takes CRAN a day or two to build the Windows binaries, so you will get a message about there not being a new version on CRAN until that happens if you're on Windows.

**Comment by yw6010** [comment]
Hi joshuaulrich, I got the same warning message when I call the tq_get() command. Is it due to the same issue? How should I fix this problem quickly? Should I use the same code provided above? Very appreciate!

**Comment by zlfang00** [comment]
@joshuaulrich 
Got errors when trying to install it to a local lib path.
The error says :ERROR: dependencies ‘xts’, ‘TTR’ are not available for package ‘quantmod’, but strangely they seems installed successfully and it does not complain about zoo is required the same way

> libpath = "/opt/zfang/R/lib/"
> remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session",lib=libpath)
Downloading GitHub repo joshuaulrich/quantmod@358-getsymbols-new.session
These packages have more recent versions available.
It is recommended to update all of them.
Which would you like to update?

1: All                         
2: CRAN packages only          
3: None                        
4: zoo (1.8-9 -> 1.8-10) [CRAN]

Enter one or more numbers, or an empty line to skip updates: 1
zoo (1.8-9 -> 1.8-10) [CRAN]
xts (NA    -> 0.12.1) [CRAN]
TTR (NA    -> 0.24.3) [CRAN]
Installing 3 packages: zoo, xts, TTR
trying URL 'https://cran.rstudio.com/src/contrib/zoo_1.8-10.tar.gz'
Content type 'application/x-gzip' length 808943 bytes (789 KB)
==================================================
downloaded 789 KB

trying URL 'https://cran.rstudio.com/src/contrib/xts_0.12.1.tar.gz'
Content type 'application/x-gzip' length 517875 bytes (505 KB)
==================================================
downloaded 505 KB

trying URL 'https://cran.rstudio.com/src/contrib/TTR_0.24.3.tar.gz'
Content type 'application/x-gzip' length 317338 bytes (309 KB)
==================================================
downloaded 309 KB

* installing *source* package ‘zoo’ ...
** package ‘zoo’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include  -I/usr/local/include  -fpic  -g -O2  -c coredata.c -o coredata.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include  -I/usr/local/include  -fpic  -g -O2  -c init.c -o init.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include  -I/usr/local/include  -fpic  -g -O2  -c lag.c -o lag.o
gcc -shared -L/opt/R360/lib64/R/lib -L/usr/local/lib64 -o zoo.so coredata.o init.o lag.o -L/opt/R360/lib64/R/lib -lR
installing to /opt/zfang/R/lib/00LOCK-zoo/00new/zoo/libs
** R
** demo
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (zoo)
* installing *source* package ‘xts’ ...
** package ‘xts’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c add_class.c -o add_class.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c any.c -o any.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c attr.c -o attr.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c binsearch.c -o binsearch.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c coredata.c -o coredata.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c diff.c -o diff.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c dimnames.c -o dimnames.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c endpoints.c -o endpoints.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c extract_col.c -o extract_col.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c init.c -o init.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c isOrdered.c -o isOrdered.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c isXts.c -o isXts.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c merge.c -o merge.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c na.c -o na.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c period_apply.c -o period_apply.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c period_arithmetic.c -o period_arithmetic.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c period_quantile.c -o period_quantile.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c rbind.c -o rbind.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c rollfun.c -o rollfun.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c runSum.c -o runSum.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c startofyear.c -o startofyear.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c subset.c -o subset.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c subset.old.c -o subset.old.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c toperiod.c -o toperiod.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c totalcols.c -o totalcols.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c tryXts.c -o tryXts.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG -I../inst/include -I"/opt/zfang/R/lib/zoo/include" -I/usr/local/include  -fpic  -g -O2  -c unique.time.c -o unique.time.o
gcc -shared -L/opt/R360/lib64/R/lib -L/usr/local/lib64 -o xts.so add_class.o any.o attr.o binsearch.o coredata.o diff.o dimnames.o endpoints.o extract_col.o init.o isOrdered.o isXts.o merge.o na.o period_apply.o period_arithmetic.o period_quantile.o rbind.o rollfun.o runSum.o startofyear.o subset.o subset.old.o toperiod.o totalcols.o tryXts.o unique.time.o -L/opt/R360/lib64/R/lib -lR
installing to /opt/zfang/R/lib/00LOCK-xts/00new/xts/libs
** R
** data
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (xts)
* installing *source* package ‘TTR’ ...
** package ‘TTR’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c adjRatios.c -o adjRatios.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c aroon.c -o aroon.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c init.c -o init.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c moving_averages.c -o moving_averages.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c percent_rank.c -o percent_rank.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c runfun.c -o runfun.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c sar.c -o sar.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c wilderSum.c -o wilderSum.o
gcc -I"/opt/R360/lib64/R/include" -DNDEBUG  -I"/opt/zfang/R/lib/xts/include" -I/usr/local/include  -fpic  -g -O2  -c zigzag.c -o zigzag.o
gcc -shared -L/opt/R360/lib64/R/lib -L/usr/local/lib64 -o TTR.so adjRatios.o aroon.o init.o moving_averages.o percent_rank.o runfun.o sar.o wilderSum.o zigzag.o -L/opt/R360/lib64/R/lib -lR
installing to /opt/zfang/R/lib/00LOCK-TTR/00new/TTR/libs
** R
** data
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (TTR)

The downloaded source packages are in
	‘/tmp/RtmpbB9zsn/downloaded_packages’
✓  checking for file ‘/tmp/RtmpbB9zsn/remotes4d43677dd4d4/joshuaulrich-quantmod-7fc8ef2/DESCRIPTION’ ...
─  preparing ‘quantmod’:
✓  checking DESCRIPTION meta-information ...
─  installing the package to process help pages
         -----------------------------------
   ERROR: dependencies ‘xts’, ‘TTR’ are not available for package ‘quantmod’
─  removing ‘/tmp/RtmpJP1VFa/Rinst50622bb4a0e0/quantmod’
         -----------------------------------
   ERROR: package installation failed
Error: Failed to install 'quantmod' from GitHub:
  System command 'R' failed, exit status: 1, stdout + stderr:
E> * checking for file ‘/tmp/RtmpbB9zsn/remotes4d43677dd4d4/joshuaulrich-quantmod-7fc8ef2/DESCRIPTION’ ... OK
E> * preparing ‘quantmod’:
E> * checking DESCRIPTION meta-information ... OK
E> * installing the package to process help pages
E>       -----------------------------------
E> ERROR: dependencies ‘xts’, ‘TTR’ are not available for package ‘quantmod’
E> * removing ‘/tmp/RtmpJP1VFa/Rinst50622bb4a0e0/quantmod’
E>       -----------------------------------
E> ERROR: package installation failed

**Comment by joshuaulrich** [comment]
> I got the same warning message when I call the tq_get() command. Is it due to the same issue? How should I fix this problem quickly? Should I use the same code provided above?

@yw6010 yes, same issue because `tq_get()` uses quantmod. Use the same call to `remotes::install_github()` provided previously.

**Comment by yw6010** [comment]
> > I got the same warning message when I call the tq_get() command. Is it due to the same issue? How should I fix this problem quickly? Should I use the same code provided above?
> 
> @yw6010 yes, same issue because `tq_get()` uses quantmod. Use the same call to `remotes::install_github()` provided previously.

I might be wrong, but I think tq_get() is under tidyquant package, like
 tq_get("DOGE-USD",
         from = start_date,
         to = until_date,
         get = "stock.prices")
Also, I receieved another warning after I run the remotes::install_github() which is 'package ‘quantmod’ is in use and will not be installed'. Thanks for your help

**Comment by joshuaulrich** [comment]
Hi @zlfang00. That's very odd and it's not clear why `remotes::install_github()` thinks the packages it just installed are not available. I could be missing something, or it may be a bug in the function. Either way, I'm afraid I can't help.

**Comment by joshuaulrich** [comment]
> 'package ‘quantmod’ is in use and will not be installed

That means you have an R session that already has quantmod loaded, so you can't change anything in the directory the package is installed in. You need to close all your R sessions on that computer, install quantmod, and then you can load and use quantmod.

**Comment by ethanbsmith** [comment]
rstudio can load and keep links to packages active for intellisense parsing and such, even after a Session -> Restart R. best to shut down everything and do installs from a clean vanilla r session.

**Comment by yw6010** [comment]
> > 'package ‘quantmod’ is in use and will not be installed
> 
> @yw6010, that means you have an R session that already has quantmod loaded, so you can't change anything in the directory the package is installed in. You need to close all your R sessions on that computer, install quantmod, and then you can load and use quantmod.


> > 'package ‘quantmod’ is in use and will not be installed
> 
> @yw6010, that means you have an R session that already has quantmod loaded, so you can't change anything in the directory the package is installed in. You need to close all your R sessions on that computer, install quantmod, and then you can load and use quantmod.

it works now!!! very appreciate. time to resume my project haha

**Comment by rhamo** [comment]
Thank you for the quick fix @joshuaulrich 
Once installing  "remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")"
this worked for me once when requesting a small group of tickers, however all new attempts since I now constantly receive "Not Cached - Error in download.." errors.

Example:

`EMN | yahoo (161|506) | Found cache file - Got 100% of valid prices | Feels good!
EMR | yahoo (162|506) | Found cache file - Got 100% of valid prices | Got it!
ENPH | yahoo (163|506) | Found cache file - Got 100% of valid prices | Boa!
EOG | yahoo (164|506) | Not Cached - Error in download..
EPAM | yahoo (165|506) | Not Cached - Error in download..
EQIX | yahoo (166|506) | Not Cached - Error in download..
EQR | yahoo (167|506) | Not Cached - Error in download..
ES | yahoo (168|506) | Not Cached - Error in download..`

I have tried restarting R but now I cannot even receive one successful return. Each ticker states "Error in download" once again. Any ideas?

Thank you

**Comment by edwinhung** [comment]
Hi, Thank you very much for the fix and comments here. getSymols.yahoo() now works for me, but I have a different problem. When I run getSymbols.yahoo() successfully in a loop for more than about 300-400 tickers, I started to get "HTTP error 401" for all following downloads. There are some failed downloads of invalid tickers in between though. Does anyone know what the issue is?

Here is an example of the subsequence download:
`> getSymbols.yahoo("AAPL",from="2022-04-27",to="2022-04-29",auto.assign=FALSE)
Warning: AAPL download failed; trying again.
Warning: Unable to import “AAPL”.
AAPL download failed after two attempts. Error message:
HTTP error 401.
[1] "Error in open.connection(file, \"rt\") : HTTP error 401.\n"
attr(,"class")
[1] "try-error"
attr(,"condition")
<simpleError in open.connection(file, "rt"): HTTP error 401.>`

Thank you!

**Comment by zpio** [comment]
install.packages("quantmod")
library(quantmod)
remotes::install_github("joshuaulrich/xts", force = T)

```
* installing *source* package 'xts' ...
ERROR: cannot remove earlier installation, is it in use?
* removing 'C:/Users/Tama/Documents/R/win-library/3.6/xts'
* restoring previous 'C:/Users/Tama/Documents/R/win-library/3.6/xts'
Error in file.copy(lp, dirname(pkgdir), recursive = TRUE, copy.date = TRUE) : 
  (converted from warning) problem copying C:\Users\Tama\Documents\R\win-library\3.6\00LOCK-xts\xts\libs\x64\xts.dll to C:\Users\Tama\Documents\R\win-library\3.6\xts\libs\x64\xts.dll: Permission denied
* removing 'C:/Users/Tama/Documents/R/win-library/3.6/xts'
* restoring previous 'C:/Users/Tama/Documents/R/win-library/3.6/xts'
Error in file.copy(lp, dirname(pkgdir), recursive = TRUE, copy.date = TRUE) : 
  (converted from warning) problem copying C:\Users\Tama\Documents\R\win-library\3.6\00LOCK-xts\xts\libs\x64\xts.dll to C:\Users\Tama\Documents\R\win-library\3.6\xts\libs\x64\xts.dll: Permission denied
Execution halted
Error: Failed to install 'xts' from GitHub:
  (converted from warning) installation of package ‘C:/Users/Tama/AppData/Local/Temp/RtmpAnQpkC/file449050346c2/xts_0.12.1.2.tar.gz’ had non-zero exit status
```

remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")

```
 installing *source* package 'quantmod' ...
   ** using staged installation
   ** R
   ** demo
   ** byte-compile and prepare package for lazy loading
   Error: package 'xts' required by 'quantmod' could not be found
   Execution halted
   ERROR: lazy loading failed for package 'quantmod'
-  removing 'C:/Users/Tama/AppData/Local/Temp/RtmpqubKOV/Rinst31ac26a1569f/quantmod'
         -----------------------------------
   ERROR: package installation failed
Error: Failed to install 'quantmod' from GitHub:
  System command 'Rcmd.exe' failed, exit status: 1, stdout + stderr (last 10 lines):
E> ** using staged installation
E> ** R
E> ** demo
E> ** byte-compile and prepare package for lazy loading
E> Error: package 'xts' required by 'quantmod' could not be found
E> Execution halted
E> ERROR: lazy loading failed for package 'quantmod'
E> * removing 'C:/Users/Tama/AppData/Local/Temp/RtmpqubKOV/Rinst31ac26a1569f/quantmod'
E>       -----------------------------------
E> ERROR: package installation failed
```
I don't know what else to do. Help

**Comment by BathosEconTrade** [comment]
get this as well when apply patch:
"Warning: package ‘quantmod’ is in use and will not be installed"
this after I removed  quantmod and Quandl and then new session then new install.
get this:
Error in new.session() : Could not establish session after 5 attempts.Error in new.session() : Could not establish session after 5 attempts.
Error in new.session() : Could not establish session after 5 attempts.
I am so grateful for quantmod and really dependent upon it

**Comment by rpfreitasxyz** [comment]
> Hi, Thank you very much for the fix and comments here. getSymols.yahoo() now works for me, but I have a different problem. When I run getSymbols.yahoo() successfully in a loop for more than about 300-400 tickers, I started to get "HTTP error 401" for all following downloads. There are some failed downloads of invalid tickers in between though. Does anyone know what the issue is? Maybe I have the same problem with @rhamo.
> 
> Here is an example of the subsequent download: `> getSymbols.yahoo("AAPL",from="2022-04-27",to="2022-04-29",auto.assign=FALSE)` `Warning: AAPL download failed; trying again. Warning: Unable to import “AAPL”. AAPL download failed after two attempts. Error message: HTTP error 401. [1] "Error in open.connection(file, \"rt\") : HTTP error 401.\n" attr(,"class") [1] "try-error" attr(,"condition") <simpleError in open.connection(file, "rt"): HTTP error 401.>`
> 
> Thank you!

I've been having the same issue!

**Comment by zlfang00** [comment]
Hi, @joshuaulrich, One thing I noticed is zoo is already installed in the public R library, while xts and TTR are not. My guess is this version quantmod only searches for its dependencies in the public library path?
Could you please confirm of that? Thanks very much!


> Hi @zlfang00. That's very odd and it's not clear why `remotes::install_github()` thinks the packages it just installed are not available. I could be missing something, or it may be a bug in the function. Either way, I'm afraid I can't help.

**Comment by BathosEconTrade** [comment]
ok   remote is working 
then use getSymbols.yahoo("AAPL)
and still get:
Error in new.session() : Could not establish session after 5 attempts.
[rpfreitasxyz](https://github.com/rpfreitasxyz) you are doing better than I at least working but this this other problem.
any insights would be great.
all my models are based on getSymbols and yahoo, and am sure everyone notices is a very interesting time now.
many thanks

**Comment by joshuaulrich** [comment]
Good news: the source package for [quantmod 0.4.20 has been released to CRAN](https://cran.r-project.org/web/packages/quantmod/)! It may take another day or two for CRAN to build the Windows and macOS binaries, so users on those systems can install from source or wait for the binaries to be built.

----

I'd like to take this opportunity to ask you to [consider sponsoring my open source work](https://github.com/sponsors/joshuaulrich/) on quantmod, xts, TTR, IBrokers, microbenchmark, and more. Tiers start at $1, and higher tiers give you access to an exclusive Slack community of my sponsors, private messages with me, and more.

----

I also want to mention an alternative to Yahoo Finance data. I've found the end of day data from [tiingo](https://api.tiingo.com/products/end-of-day-stock-price-data) to be higher quality and the endpoint is more stable than Yahoo Finance. You need to sign up to get a free API key, and that gives individuals free access with 500 requests/hour and 20,000 requests/day, up to 5GB of data/month. (Disclaimer: While I don't receive any referral compensation for pointing individuals to the Free Plan or Power Plan, I may receive compensation--I honestly don't remember how much--from them for new Commercial Plan customers who learned of tiingo through me. But you have to mention me when you join, I don't have an affiliate link or anything like that).

**Comment by joshuaulrich** [comment]
@rhamo, @edwinhung, and @rpfreitasxyz A HTTP 401 response means the request is "unauthorized". That can happen for various reasons, but my guess is that you're hitting rate limits imposed by the Yahoo Finance servers. `getSymbols()` adds pauses when you request a lot of symbols at once to try and avoid hitting these rate limits. @msperlin (author of BatchGetSymbols) may have a better idea.

----

@zpio you shouldn't need to install xts from GitHub. You only need the version on CRAN. Can you try to install quantmod via `remotes::install_github("joshuaulrich/quantmod")` and see if that works?

----

@zlfang00 The new version of quantmod uses zoo, xts, and TTR just like prior versions. Nothing has changed in that regard, so I'm stumped about what's causing that result.

----

> Error in new.session() : Could not establish session after 5 attempts.

@BathosEconTrade That looks like version 0.4.18 is being loaded instead of 0.4.19 or 0.4.20. Check `packageVersion("quantmod")` to see which version you're loading. Double check your library paths in case you are using more than one library...

**Comment by BathosEconTrade** [comment]
Many thanks all!   Great help.  Must have version 4.2.

**Comment by rhamo** [comment]
> Thank you for the quick fix @joshuaulrich Once installing "remotes::install_github("joshuaulrich/quantmod@358-getsymbols-new.session")" this worked for me once when requesting a small group of tickers, however all new attempts since I now constantly receive "Not Cached - Error in download.." errors.
> 
> Example:
> 
> EMN | yahoo (161|506) | Found cache file - Got 100% of valid prices | Feels good! EMR | yahoo (162|506) | Found cache file - Got 100% of valid prices | Got it! ENPH | yahoo (163|506) | Found cache file - Got 100% of valid prices | Boa! EOG | yahoo (164|506) | Not Cached - Error in download.. EPAM | yahoo (165|506) | Not Cached - Error in download.. EQIX | yahoo (166|506) | Not Cached - Error in download.. EQR | yahoo (167|506) | Not Cached - Error in download.. ES | yahoo (168|506) | Not Cached - Error in download..
> 
> I have tried restarting R but now I cannot even receive one successful return. Each ticker states "Error in download" once again. Any ideas?
> 
> Thank you

Have tried with new quantmod 0.4.20 update but still no success. Every ticker states "Not Cached - Error in download" error as above.

Any ideas on how to solve this issue?
Thanks again

**Comment by samiranvekar** [comment]
> 

Thank you@joshuaulrich. The issue is resolved

**Comment by msperlin** [comment]
It seems that yahoo finance is now restricting access to the data. Once you download too much data, you get the "HTTP 401" error. This is controled by Yahoo and there is nothing myself or @joshuaulrich can do about it.

As with any webscrapping script, you can try to overcome the restriction with:
- changing your ip one you reach the limit (use VPN)
- use a caching system to save data locally and download data over the course of days. Usually the restricted ips are released in 24 hours. A caching system is already available in [yfR](https://github.com/msperlin/yfR)/BatchGetSymbols).

Overall, if the limits are too restrictive, it might be best to find other source of price data.

**Comment by NilssonHedge** [comment]
Is there any indication of what the Yahoo Rate limit is?

**Comment by msperlin** [comment]
@NilssonHedge Im looking into it right now.. It doenst seem too restrictive. After a 401 error, I can download again in a few minutes.

**Comment by zlfang00** [comment]
> 
are you sure it is the yahoo restriction? I tried Python package yfinance to grab the same size of date, no issues.

**Comment by joshuaulrich** [comment]
> are you sure it is the yahoo restriction? I tried Python package yfinance to grab the same size of date, no issues.

The `yfinance` python library uses a different endpoint that returns JSON instead of a CSV. I'll look into adding something to `getSymbols()` that tries the other endpoint if one of them fails.

**Comment by rhamo** [comment]
Changing ip has worked wonders. Once again BatchGetSymbols now retrieves data without errors. 

Thanks so much for all your help @msperlin and @joshuaulrich

**Comment by zpio** [comment]
> 
thanks, it works now

**Comment by mif2000** [comment]
Hi,  Really appreciate your dedication.  I'm a new user to R (v 3.5.3) through Bert for Excel.  I finally had everything working to call R functionality and getSymbols (with a single ticker) through an Excel VBA macro and then it abruptly stopped as noted in prior comments.  It appears that the problem with getSymbols has been resolved and that I can now download quantmod v 4.20 zip file for Windows (10), which I did from the quantmod.zip link on the CRAN page dtd 2022-04-29, copying the unzipped quantmod directory to the BERT2/lib folder. The command "library (quantmod)" ouputs an error: "package or namespace load failed for 'quantmod' in rbind(info, getNamespaceInfo(env,"S3methods")).  packageVersion("quantmod") outputs '0.4.20'.  I must be doing something wrong!  What is the proper way to install the new version of quantmod?

Thanks for your help,
mif2000
